lib/class-wp-block-supports.php

```diff
@@ -0,0 +1,454 @@
+					'gradient' => array(
+						'type' => 'string',
+					),
+				),
+				'callback'   => '__return_false',
+				'default'    => array( false, false ),
+			),
+			'color.link'       => array(
+				'attributes' => array(),
+				'callback'   => '__return_false',
+				'default'    => array( false, false ),
+			),
+			'color.text'       => array(
+				'attributes' => array(
+					'style'     => array(
+						'type' => 'object',
+					),
+					'textColor' => array(
+						'type' => 'string',
+					),
+				),
+				'callback'   => function( $attributes ) {
+					$has_named_text_color = array_key_exists( 'textColor', $attributes );
+					$has_custom_text_color = isset( $attributes['style']['color']['text'] );
+					$classes = '';
+					$styles  = '';
+
+					// Apply required generic class.
+					if ( $has_custom_text_color || $has_named_text_color ) {
+						$classes .= ' has-text-color';
+					}
+					// Apply color class or inline style.
+					if ( $has_named_text_color ) {
+						$classes .= sprintf( ' has-%s-color', $attributes['textColor'] );
+					} elseif ( $has_custom_text_color ) {
+						$styles = sprintf(
+							'color:%s;',
+							$attributes['style']['color']['text']
+						);
+					}
+					return array(
+						'class' => $classes,
+						'style' => $styles,
+					);
+				},
+				'default'    => array( false, true ),
+			),
+			'customClassName'  => array(
+				'attributes' => array(
+					'className' => array(
+						'type' => 'string',
+					),
+				),
+				'callback'   => function( $attributes ) {
+					return array_key_exists( 'className', $attributes ) ?
+						array( 'class' => $attributes['className'] ) :
+						false;
+				},
+				'default'    => true,
+			),
+			'fontSize'         => array(
+				'attributes' => array(
+					'style'    => array(
+						'type' => 'object',
+					),
+					'fontSize' => array(
+						'type' => 'string',
+					),
+				),
+				'callback'   => function( $attributes, $block_name ) {
+					die( 42 );
+					$has_named_font_size = array_key_exists( 'fontSize', $attributes );
+					$has_custom_font_size = isset( $attributes['style']['typography']['fontSize'] );
+
+					$classes = '';
+					$styles  = '';
+					if ( $has_named_font_size ) {
+						$classes = sprintf( 'has-%s-font-size', $attributes['fontSize'] );
+					} elseif ( $has_custom_font_size ) {
+						$styles = sprintf(
+							'font-size:%spx;',
+							$attributes['style']['typography']['fontSize']
+						);
+					}
+
+					return array(
+						'class' => $classes,
+						'style' => 'color:pink;', //$styles,
+					);
+				},
+				'default'    => false,
+			),
+			'lineHeight'       => array(
+				'attributes' => array(
+					'style' => array(
+						'type' => 'object',
+					),
+				),
+				'callback'   => function( $attributes, $block_name ) {
+					$has_line_height = isset( $attributes['style']['typography']['lineHeight'] );
+
+					$classes = '';
+					if ( $has_line_height ) {
+						$generated_class_name = uniqid( 'wp-block-lineheight-' );
```

Comment:
It seems with this PR you're also trying to change how we generate the custom styles in the server. While this is probably a sane approach, IMO, we should stick to the current way for the moment and do this in its own PR. Basically just keep the inline styles for now.

Reply:
which also mean `wp-block-supports` style shouldn't be needed yet.

Reply:
Agreed. That’s a remnant from when into class was supported.

Reply:
As an aside if we try to generate inline style on the server. I think generating a class name per block instance and then add the styles of that block to that class name is not ideal.

I think we would have the opportunity to use atomic CSS ( each class/selector contains just one rule). Per CSS style rule we e.g: "line-height: 5px" we would have a class e.g: ".Ab { line-height: 5px }". Each block that has a line-height 5px would simply refer to that class. Our preset classes are already atomic so in the case of presets instead of generating a random class, we would simply refer to a human-readable class.
Implementing this is not complex we would have a global dictonaty that maps styles to classes e.g:
```
array(
	'line-height' => array(
		'5px' => 'ab'
	)
);
```
and a counter:
```
$current_class_count = 3;
```
Each time a style has a class we would simply include that class. If no class was defined we would encode the counter in Base36 (uses A-Z and 0-9) update the mapper to map to that class and increment the counter.


Atomic CSS is deployed in production at Facebook and Twitter and seems to reduce the amount of CSS required.